native/cocos/renderer/gfx-base/memory/TLSFPool.h

```diff
@@ -0,0 +1,131 @@
+/****************************************************************************
+Copyright (c) 2019-2022 Xiamen Yaji Software Co., Ltd.
+
+http://www.cocos.com
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated engine source code (the "Software"), a limited,
+worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+to use Cocos Creator solely to develop games on your target platforms. You shall
+not use Cocos Creator software for developing other software or tools that's
+used for developing games. You are not granted to publish, distribute,
+sublicense, and/or sell copies of Cocos Creator.
+
+The software or tools in this License Agreement are licensed, not sold.
+Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+****************************************************************************/
+
+#pragma once
+
+#include "gfx-base/memory/ObjectChunkPool.h"
+#include "gfx-base/memory/MemoryView.h"
+
+namespace cc::gfx {
+/**
+ * @en constexpr version of log2
+ * @tparam T Data type, must be integral
+ * @param n Input value, must be power of 2
+ * @return log2 result of n
+ */
+template <typename T>
+constexpr T log2(T n) { // NOLINT(misc-no-recursion)
+    static_assert(std::is_integral_v<T>, "T must be integral");
+    return ((n < 2) ? 0 : 1 + log2(n / 2));
+}
+
+struct TLSFPoolInfo {
+    uint64_t poolSize = 0; // if poolSize equals to 0, using default pool size TLSFPool::DEFAULT_POOL_SIZE
+};
+
+/**
+ * Dynamic Memory Allocator (Two-Level Segregated Fit memory allocator) for Video Memory.
+ * Original paper: http://www.gii.upv.es/tlsf/files/ecrts04_tlsf.pdf
+ */
+class TLSFPool {
+public:
+    TLSFPool() = default;
+    ~TLSFPool() = default;
+
+    TLSFPool(const TLSFPool &) = delete;
+    TLSFPool &operator=(const TLSFPool &) = delete;
+
+    TLSFPool(TLSFPool &&) noexcept = default;
+    TLSFPool &operator=(TLSFPool &&) noexcept = default;
+
+    /**
+     * @en Block info of an allocation
+     */
+    struct Block : public MemoryView {
+        Block *prevPhy = nullptr;
+        Block *nextPhy = nullptr;
+
+        Block *prevFree = nullptr;
+        Block *nextFree = nullptr;
+
+        bool isFree() const { return prevFree != this; }
+    };
+
+    /**
+     * @en Initialize the pool. Should be called before allocation.
+     * @param poolInfo Pool create info.
+     */
+    void initialize(const TLSFPoolInfo &poolInfo);
```

Comment:
Why need it? Can just pass info to constructor?

Reply:
I think it is better to use initialize if operation may failed in constructor, e.g. `allocation`, `vector::resize`.

Reply:
But it returns void. Then what's the difference from causing error in constructor?

Reply:
And as it is a public function. Then can invoke it many times. Does it support this usage?

Reply:
Mainly to avoid exceptions in constructors, since allocation usually come with exceptions.

Reply:
exception happens in initialize also have problem, as don't do any protect. So i still don't know what's the benefit of it.

Reply:
Got it.